engine/main.ml

```diff
@@ -224,39 +230,337 @@ let rec parse_lambda lsp =
+  in
+  let print_env env =
+    let buf = Buffer.create 42 in
+    bprint_env 0 buf env;
+    BatIO.write_line BatIO.stdout (Buffer.contents buf)
+  in
+  let expand_env variable entry : environment SMap.t = 
+    let _ = match SMap.find_opt variable env with
+      | Some _ -> assert false
+      | None -> ()
+    in
+    SMap.add variable entry env
+  in
+  (* perform union on two maps, keys should never differ *)
+  let union env1 env2 = SMap.union (fun _ a b -> assert (a = b); Some a) env1 env2
+  in
+  let match_let_accessor env acc =
+    match acc with
+    | Var v -> Accessor (AcRoot v)
+    | Field (i, v) ->
+      let acc = match SMap.find_opt v env with
+        | Some (Accessor a) -> a
+        | Some (Addition _) -> assert false
+        | Some (Function _) -> assert false
+        | Some (Catch _) -> assert false
+        | None -> assert false
+      in
+      Accessor (AcField (acc, i))
+    | Function (v, sxp) ->
+      let env' = expand_env v (Accessor (AcRoot v))
+      in
+      let constraint_tree = sym_exec sxp constraints env' in
+      Function (v, constraint_tree)
+    | _ -> assert false
+  in
+  match sexpr with
+  | Let (blist, next_sexpr) ->
+    let env' = blist |>
+               List.map (fun (var, sxp) -> expand_env var (match_let_accessor env sxp)) |>
+               List.fold_left union env
+    in
+    sym_exec next_sexpr constraints env'
+  | If (bexpr, strue, sfalse) ->
+    let piop, sxp =
+      match bexpr with
+      | Comparison (bop, sxp, i) -> (match_bop (bop, i)), sxp
+      | Isout (i, v) -> Isout i, Var v
+      | _ -> assert false (* TODO *)
+    in
+    let avar = match sxp with
+      | Var v -> begin
+          match SMap.find_opt v env with
+        | Some (Accessor a) -> a
+        | _ -> assert false
+        end
+      | _ -> assert false
+    in
+    let child1 = sym_exec strue ({var=avar; op=piop}::constraints) env
+    in
+    let child2 = sym_exec sfalse ({var=avar; op=negate piop}::constraints) env
+    in
+    Node [child1; child2;]
+  | Switch (sxp, swlist, defcase) ->
+    let var = match sxp with
+      | Var v -> begin
+          match SMap.find_opt v env with
+        | Some (Accessor a) -> a
+        | _ -> assert false
+        end
+      | _ -> assert false
+    in
+    let constraintsxtargets = List.map (fun c -> (match_switch (fst c), snd c)) swlist
+    in
+    let defcase_constraints = List.map (fun (c,_) -> {var=var; op=negate c}) constraintsxtargets
+    in
+    let children = constraintsxtargets |>
+                   List.map (fun (c, target) -> sym_exec target ({var=var; op=c}::constraints) env)
+    in
+    begin
+      match defcase with
+      | Some defcase -> Node (children@[sym_exec defcase (defcase_constraints@constraints) env])
+      | None -> Node children
+    end
+  | Catch (sxp, extpt, varlist, sxp') ->
+    let c_tree = sym_exec sxp' constraints env in
+    let env' = expand_env (string_of_int extpt) (Catch (extpt, varlist, c_tree)) in
```

Comment:
the `string_of_int` here is really ugly, we should fix it after this PR

Reply:
(by turning the environment into a record: variable->value environment, exitpoint->constraint-tree environment.)